### PR TITLE
#286: Support for modal-like image viewing and zooming

### PR DESCRIPTION
**Pull request by beneditomauro**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-components-react-native/issues/286 |
| Decisions | Double-tap zoom <br> Pinch zoom <br> Translate to the "edge" while zoomed in <br>|
| Animated GIF | <br> IOS: <br> 2P5v4m5Kna <br> Android: <br> HfaiLkN0I1|

### CODE REVIEW

File: react/components/atoms/lightbox/lightbox.js [hunk 1]
```diff
@@ -39,8 +54,36 @@ export class Lightbox extends PureComponent {
...
     constructor(props) {
         super(props);
+
+        this.screenWidth = Dimensions.get("window").width;
+        this.screenHeight = Dimensions.get("window").height;
+
+        this.fullscreenContainerBackgroundColor = 0;
+        this.fullscreenContainerBackgroundColorAnimated = 1;
+        this.fullscreenContainerTranslateYAnimationValue = 15;
+        this.fullscreenContainerBackgroundColorAnimatedDuration = 200;
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:64]: dead code?

File: react/components/atoms/lightbox/lightbox.js [hunk 2]
```diff
@@ -39,8 +54,36 @@ export class Lightbox extends PureComponent {
...
+        this.scaleRate = new Animated.Value(1);
+        this.lastScale = 1;
+
         this.state = {
-            visible: false
+            imageSizes: null,
+            zoomed: false,
+            zooming: false,
+            visible: this.props.visible,
+            fullscreenContainerBackgroundColor: new Animated.Value(0),
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:83]: perhaps only `containerBackgroundColor` or `backgroundColor` since there is no other background color being set

File: react/components/atoms/lightbox/lightbox.js [hunk 3]
```diff
@@ -39,8 +54,36 @@ export class Lightbox extends PureComponent {
...
+        this.translatedY = 0;
+        this.translatedYTreshold = 0;
+
+        this.baseScale = new Animated.Value(1);
+        this.scaleRate = new Animated.Value(1);
+        this.lastScale = 1;
+
         this.state = {
-            visible: false
+            imageSizes: null,
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:79]: dead code

File: react/components/atoms/lightbox/lightbox.js [hunk 4]
```diff
@@ -65,14 +109,184 @@ export class Lightbox extends PureComponent {
...
+            })
+        ]).start();
+    };
+
     onLightboxPress = () => {
         this.setVisibility(true);
     };
 
+    onPanGesture = event => {
+        if (this.state.scale._value === 1 || !this.state.zoomed) return;
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:141]: newline below

File: react/components/atoms/lightbox/lightbox.js [hunk 5]
```diff
@@ -65,14 +109,184 @@ export class Lightbox extends PureComponent {
...
+        const dx = event.nativeEvent.translationX + this.translatedX;
+        const dy = event.nativeEvent.translationY + this.translatedY;
+        if (Math.abs(dx) < this.translatedXTreshold) {
+            this.state.translateX.setValue(dx);
+        }
+
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:147]: remove newline

File: react/components/atoms/lightbox/lightbox.js [hunk 6]
```diff
@@ -65,14 +109,184 @@ export class Lightbox extends PureComponent {
...
+        if (Math.abs(dy) < this.translatedYTreshold) {
+            this.state.translateY.setValue(dy);
+        }
+    };
+
+    onPanGestureEnd = event => {
+        if (event.nativeEvent.state === State.END) {
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:154]: perhaps early exit with `if (event.nativeEvent.state !== State.END) return;`

File: react/components/atoms/lightbox/lightbox.js [hunk 7]
```diff
@@ -65,14 +109,184 @@ export class Lightbox extends PureComponent {
...
+        const pinchScale = event.nativeEvent.scale;
+        const scale = this.baseScale._value * pinchScale;
+        if (scale < 1 || scale > 2) return;
+        this.scaleRate.setValue(pinchScale);
+        this.lastScale = scale;
+        this.setState({ zoomed: true });
+    };
+
+    onPinchGestureEnd = event => {
+        if (event.nativeEvent.oldState === State.ACTIVE) {
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:170]: perhaps early exit with `if (event.nativeEvent.oldState !== State.ACTIVE) return;`

File: react/components/atoms/lightbox/lightbox.js [hunk 8]
```diff
@@ -65,14 +109,184 @@ export class Lightbox extends PureComponent {
...
+            this.translatedX = this.state.translateX._value;
+            this.translatedY = this.state.translateY._value;
+        }
+    };
+
+    onPinchGesture = event => {
+        const pinchScale = event.nativeEvent.scale;
+        const scale = this.baseScale._value * pinchScale;
+        if (scale < 1 || scale > 2) return;
+        this.scaleRate.setValue(pinchScale);
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:164]: newline above

File: react/components/atoms/lightbox/lightbox.js [hunk 9]
```diff
@@ -65,14 +109,184 @@ export class Lightbox extends PureComponent {
...
+                this.setState({ zooming: false, zoomed: false });
+            });
+        });
+    };
+
     _imageSource = () => {
         return this.props.uri ? { uri: this.props.uri } : this.props.src;
     };
 
+    _sanatizeFloat = x => {
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:262]: why is this required? Add comments explaining the need for this logic

File: react/components/atoms/lightbox/lightbox.js [hunk 10]
```diff
@@ -106,20 +327,43 @@ export class Lightbox extends PureComponent {
...
-                            style={styles.buttonClose}
-                            iconStrokeWidth={2}
-                            size={isTabletSize() ? 52 : 34}
-                            iconHeight={isTabletSize() ? 34 : 22}
-                            iconWidth={isTabletSize() ? 34 : 22}
-                            backgroundColor={"#000000"}
-                            iconStrokeColor={"#ffffff"}
-                        />
-                    </View>
+                    <GestureHandlerRootView style={{ width: "100%", height: "100%" }}>
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:330]: move style to stylesheet below

**Review by BeeMargarida — COMMENTED** [review]
Just a first review since I was not yet able to test the feature on my phone. Will try again later in the office

File: react/components/atoms/lightbox/lightbox.js [hunk 11]
```diff
@@ -89,7 +293,14 @@ export class Lightbox extends PureComponent {
         return [
             styles.imageFullscreen,
             {
-                resizeMode: this.props.resizeModeFullScreen
+                width: this.screenWidth,
+                resizeMode: "contain",
+                transform: [
+                    { perspective: 200 },
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:299]: what does this do in this scenario?

File: react/components/atoms/lightbox/lightbox.js [hunk 12]
```diff
@@ -89,7 +293,14 @@ export class Lightbox extends PureComponent {
         return [
             styles.imageFullscreen,
             {
-                resizeMode: this.props.resizeModeFullScreen
+                width: this.screenWidth,
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:296]: isn't it just enough to do `width: "100%"`?

File: react/components/atoms/lightbox/lightbox.js [hunk 13]
```diff
@@ -106,20 +317,45 @@ export class Lightbox extends PureComponent {
...
-                            style={styles.buttonClose}
-                            iconStrokeWidth={2}
-                            size={isTabletSize() ? 52 : 34}
-                            iconHeight={isTabletSize() ? 34 : 22}
-                            iconWidth={isTabletSize() ? 34 : 22}
-                            backgroundColor={"#000000"}
-                            iconStrokeColor={"#ffffff"}
-                        />
-                    </View>
+                    <GestureHandlerRootView style={styles.gestureHandlerRootView}>
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:320]: Regarding what they say in their documentation, this should wrap the entire library code or we need to put it in the Robin's and other apps `app.js` https://docs.swmansion.com/react-native-gesture-handler/docs/#js

File: react/components/atoms/lightbox/lightbox.js [hunk 14]
```diff
@@ -106,20 +317,45 @@ export class Lightbox extends PureComponent {
...
+                                <TapGestureHandler
+                                    onHandlerStateChange={this.onDoubleTap}
+                                    numberOfTaps={2}
+                                >
+                                    <PinchGestureHandler
+                                        onGestureEvent={this.onPinchGesture}
+                                        onHandlerStateChange={this.onPinchGestureEnd}
+                                    >
+                                        <Animated.Image
+                                            resizeMode={"contain"}
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:337]: similar to above, should it not use the resizeMode in props

File: react/components/atoms/lightbox/lightbox.js [hunk 15]
```diff
@@ -69,10 +131,152 @@ export class Lightbox extends PureComponent {
...
+    _getTranslateX = event => {
+        const x = event.x;
+        const middle = this.screenWidth / 2;
+        const touchMiddleDistance = x - middle;
+
+        const translateX =
+            Math.abs(touchMiddleDistance) > this.translatedXTreshold
+                ? this.translatedXTreshold
+                : touchMiddleDistance;
+        return x > middle ? -1 * Math.abs(translateX) : Math.abs(translateX);
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:277]: Perhaps simplify with the code below and also add comments about the logic behind this function
```suggestion
        const x = event.x;
        const middle = this.screenWidth / 2;
        const touchMiddleDistance = x - middle;
        const translateX = Math.min(this.translatedXTreshold, Math.abs(touchMiddleDistance))
        return x > middle ? -1 * translateX : translateX;

```

File: react/components/atoms/lightbox/lightbox.js [hunk 16]
```diff
@@ -69,10 +131,152 @@ export class Lightbox extends PureComponent {
...
+    _getTranslateY = event => {
+        const y = event.y;
+        const middle = this.screenHeight / 2;
+        const touchMiddleDistance = middle - y;
+        const translateY =
+            Math.abs(touchMiddleDistance) > this.translatedYTreshold
+                ? this.translatedYTreshold
+                : touchMiddleDistance;
+
+        return y > middle ? -1 * Math.abs(translateY) : Math.abs(translateY);
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:265]: Perhaps simplify with the code below and also add comments about the logic behind this function
```suggestion
        const y = event.y;
        const middle = this.screenHeight / 2;
        const touchMiddleDistance = middle - y;
        const translateY = Math.min(this.translatedYTreshold, Math.abs(touchMiddleDistance))
        return y > middle ? -1 * translateY : translateY;
```

File: react/components/atoms/lightbox/lightbox.js [hunk 17]
```diff
@@ -69,10 +131,152 @@ export class Lightbox extends PureComponent {
...
+            } else {
+                this._doubleTapZoomOut();
+            }
+        }
+    };
+
+    _doubleTapZoomIn = event => {
+        const scaledImage = this.screenHeight * 2;
+        this.translatedXTreshold = this.screenWidth / 4;
+        this.translatedYTreshold = (scaledImage - this.screenHeight) / 4;
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:186]: add comment explaining this logic

File: react/components/atoms/lightbox/lightbox.js [hunk 18]
```diff
@@ -69,10 +131,152 @@ export class Lightbox extends PureComponent {
...
+        if (event.nativeEvent.state !== State.END) return;
+
+        this.translatedX = this.state.translateX._value;
+        this.translatedY = this.state.translateY._value;
+    };
+
+    onPinchGesture = event => {
+        const pinchScale = event.nativeEvent.scale;
+        const scale = this.baseScale._value * pinchScale;
+        if (scale < 1 || scale > 2) return;
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:157]: why these limits? For example, whats app seems to support a bigger scale.
Perhaps move the bigger value to a prop, so that it can be customizable

File: react/components/atoms/lightbox/lightbox.js [hunk 19]
```diff
@@ -69,10 +131,152 @@ export class Lightbox extends PureComponent {
...
+        this.setState({ zoomed: true });
+    };
+
+    onPinchGestureEnd = event => {
+        if (event.nativeEvent.oldState !== State.ACTIVE) return;
+        this.resetTranslation();
+        this.baseScale.setValue(this.lastScale);
+        this.scaleRate.setValue(1);
+        this.translatedXTreshold = (this.screenWidth * this.lastScale - this.screenWidth) / 4;
+        this.translatedYTreshold = (this.screenHeight * this.lastScale - this.screenHeight) / 4;
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:170]: add a comment explaining the logic behind these thresholds and the reset usage

File: react/components/atoms/lightbox/lightbox.js [hunk 20]
```diff
@@ -69,10 +131,152 @@ export class Lightbox extends PureComponent {
...
+    };
+
+    onDoubleTap = event => {
+        if (event.nativeEvent.state === State.ACTIVE) {
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:174]: early exit

**Review by BeeMargarida — CHANGES_REQUESTED** [review]
There is some weird behaviour happening in some use cases

https://user-images.githubusercontent.com/25725586/151969990-a01e4ca4-7889-40d6-993d-8bdea6585535.mp4

File: react/components/atoms/lightbox/lightbox.js [hunk 21]
```diff
@@ -39,8 +54,30 @@ export class Lightbox extends PureComponent {
...
+        this.fullZoomedInValue = 2;
+        this.fullZoomedOutValue = 1;
+
+        this.translatedX = 0;
+        this.translatedXTreshold = 0;
+        this.translatedY = 0;
+        this.translatedYTreshold = 0;
+
+        this.baseScale = new Animated.Value(1);
+        this.scaleRate = new Animated.Value(1);
```
[the remark this answers is not in the document]
Comment (beneditomauro) [reply react/components/atoms/lightbox/lightbox.js:70]: They're referenced in a propery in  state

File: react/components/atoms/lightbox/lightbox.js [hunk 22]
```diff
@@ -69,10 +137,161 @@ export class Lightbox extends PureComponent {
...
+     *
+     * @param {Object} event The double tap event
+     */
+    _getDoubleTapZoomInTranslateX = event => {
+        const x = event.x;
+        const middle = this.screenWidth / this.fullZoomedInValue;
+        const touchMiddleDistance = x - middle;
+
+        // if the distance of the coordinate is greater than the translate treshold
+        // then assume the treshold value as the X coordinate to zoom in to
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:290]: with the comments of the function, either maintain this one and remove the function one or vice-versa

File: react/components/atoms/lightbox/lightbox.js [hunk 23]
```diff
@@ -69,10 +137,161 @@ export class Lightbox extends PureComponent {
...
+                Animated.timing(this.state.translateY, {
+                    toValue: translateY,
+                    duration: this.props.translateAnimationDuration,
+                    useNativeDriver: true
+                })
+            ]).start(() => {
+                const scaleRatio = this.baseScale._value * this.fullZoomedInValue;
+                this.lastScale = scaleRatio;
+                this.baseScale.setValue(this.lastScale);
+                this.scaleRate.setValue(1);
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:222]: Is this related to that problem you told me about? Since you are setting the scaleRate above to `this.fullZoomedInValue` and later you reset it.
Either way, why are these values not being used in the transform below?

**Review by BeeMargarida — CHANGES_REQUESTED** [review]

File: react/components/atoms/lightbox/lightbox.js [hunk 24]
```diff
@@ -89,11 +308,32 @@ export class Lightbox extends PureComponent {
...
+     * determines how much space is available after translation
+     * until reach the end of the scaled "image".
+     *
+     * @param {Float} scale Current scale value of the image
+     */
+    _setTranslateTresholds = scale => {
+        const scaledHeight = this.screenHeight * scale;
+        const scaledWidth = this.screenWidth * scale;
+
+        const scaleDivisor = scale * 2;
```
[the remark this answers is not in the document]
Comment (beneditomauro) [reply react/components/atoms/lightbox/lightbox.js:332]: For some reason, the math only works when the extra space is divided by the double value of the scale, so I would say yes "it's just the number that best adjusts the image thresholds"

File: react/components/atoms/lightbox/lightbox.js (hunk 23 above)
Comment (beneditomauro) [reply react/components/atoms/lightbox/lightbox.js:222]: Yes, the redundancy is about the problem I've talked about. 
The values alter the `scale` state property

File: react/components/atoms/lightbox/lightbox.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (beneditomauro) [reply react/components/atoms/lightbox/lightbox.js:73]: Yes, the `scale` to transform to is a multiplication of these two

File: react/components/atoms/lightbox/lightbox.js [hunk 25]
```diff
@@ -106,37 +342,59 @@ export class Lightbox extends PureComponent {
...
-                            iconStrokeWidth={2}
-                            size={isTabletSize() ? 52 : 34}
-                            iconHeight={isTabletSize() ? 34 : 22}
-                            iconWidth={isTabletSize() ? 34 : 22}
-                            backgroundColor={"#000000"}
-                            iconStrokeColor={"#ffffff"}
-                        />
-                    </View>
+                    <Animated.View style={styles.fullscreenContainer}>
+                        <PanGestureHandler
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:346]: It seems the removal of `GestureHandlerRootView` causes this not to work on Android, even by adding `GestureHandlerRootView` to the app.js of robin, for example. For it to work it would require the usage of `gestureHandlerRootHOC`, like it says in the [documentation](https://docs.swmansion.com/react-native-gesture-handler/docs/#usage-with-modals-on-android). I tried and there is some flickering happening when zooming so perhaps maybe reverting this change (I'm sorry, I know I was the one to tell you to remove it). What do you think?

File: react/components/atoms/lightbox/lightbox.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (BeeMargarida) [reply react/components/atoms/lightbox/lightbox.js:73]: Nvm, I forgot that part!

File: react/components/atoms/lightbox/lightbox.js [hunk 26]
```diff
@@ -39,8 +57,29 @@ export class Lightbox extends PureComponent {
...
+        this.lastScale = 1;
+
         this.state = {
-            visible: false
+            zoomed: false,
+            zooming: false,
+            visible: this.props.visible,
+            scale: Animated.multiply(this.baseScale, this.scaleRate),
+            translateX: new Animated.Value(0),
+            translateY: new Animated.Value(0)
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:82]: You can remove these from state since they are animations, so that they can stay together with the other 2 variables mentioned above

**Review by BeeMargarida — COMMENTED** [review]
Also another weird behaviour

https://user-images.githubusercontent.com/25725586/152167390-d07eae70-021c-4972-8cc5-0771996f0663.mp4

File: react/components/atoms/lightbox/lightbox.js [hunk 27]
```diff
@@ -28,9 +41,15 @@ export class Lightbox extends PureComponent {
             src: undefined,
             width: "100%",
             height: "100%",
+            zoomAnimationDuration: 200,
+            translateAnimationDuration: 200,
+            maxZoomScale: 3,
+            minZoomScale: 1,
```
Comment (BeeMargarida) [inline react/components/atoms/lightbox/lightbox.js:47]: switch order to make it similar to the order above

**Review by BeeMargarida — COMMENTED** [review]
There a feature missing and some improvements to be made:
- [feature] The pinch zoom should start on the position of the pinch (in the video using Telegram it works like that - in WhatsApp as well)
- [improvement] There is a weird slide effect when double tapping on the edges. In WhatsApp is there as well (a bit smoother though) but in Telegram it is not. I don't know if this is worth spending a lot of time fixing, but wanted to take note of it.
- In some places the pinching does not work

This is a lot of stuff, so if you need any help please say so and I can explore this with you.

https://user-images.githubusercontent.com/25725586/152194148-1da18fba-baa5-499c-9ade-a99dda8fee22.mp4

**Comment by joamag** [comment]
@BeeMargarida @beneditomauro let's try to integrate this as soon as possible in RIPE Robin
